cpp/src/groupby/sort/aggregate.cpp

```diff
@@ -795,6 +795,65 @@ void aggregate_result_functor::operator()<aggregation::MERGE_TDIGEST>(aggregatio
+{
+  if (cache.has_result(values, agg)) { return; }
+
+  auto const& udf_ptr   = dynamic_cast<cudf::detail::host_udf_aggregation const&>(agg).udf_ptr;
+  auto const data_attrs = [&]() -> host_udf_base::data_attribute_set_t {
+    if (auto tmp = udf_ptr->get_required_data(); !tmp.empty()) { return tmp; }
+    // Empty attribute set means everything.
+    return {host_udf_base::groupby_data_attribute::INPUT_VALUES,
+            host_udf_base::groupby_data_attribute::GROUPED_VALUES,
+            host_udf_base::groupby_data_attribute::SORTED_GROUPED_VALUES,
+            host_udf_base::groupby_data_attribute::NUM_GROUPS,
+            host_udf_base::groupby_data_attribute::GROUP_OFFSETS,
+            host_udf_base::groupby_data_attribute::GROUP_LABELS};
+  }();
+
+  // Do not cache udf_input, as the actual input data may change from run to run.
+  host_udf_base::input_map_t udf_input;
+  for (auto const& attr : data_attrs) {
+    CUDF_EXPECTS(std::holds_alternative<host_udf_base::groupby_data_attribute>(attr.value) ||
+                   std::holds_alternative<std::unique_ptr<aggregation>>(attr.value),
+                 "Invalid input data attribute for HOST_UDF groupby aggregation.");
+    if (std::holds_alternative<host_udf_base::groupby_data_attribute>(attr.value)) {
+      switch (std::get<host_udf_base::groupby_data_attribute>(attr.value)) {
+        case host_udf_base::groupby_data_attribute::INPUT_VALUES:
+          udf_input.emplace(attr, values);
+          break;
+        case host_udf_base::groupby_data_attribute::GROUPED_VALUES:
+          udf_input.emplace(attr, get_grouped_values());
+          break;
+        case host_udf_base::groupby_data_attribute::SORTED_GROUPED_VALUES:
+          udf_input.emplace(attr, get_sorted_values());
+          break;
+        case host_udf_base::groupby_data_attribute::NUM_GROUPS:
+          udf_input.emplace(attr, helper.num_groups(stream));
+          break;
+        case host_udf_base::groupby_data_attribute::GROUP_OFFSETS:
+          udf_input.emplace(attr, helper.group_offsets(stream));
+          break;
+        case host_udf_base::groupby_data_attribute::GROUP_LABELS:
+          udf_input.emplace(attr, helper.group_labels(stream));
+          break;
+        default: CUDF_UNREACHABLE("Invalid input data attribute for HOST_UDF groupby aggregation.");
+      }
+    } else {  // data is result from another aggregation
+      auto other_agg = std::get<std::unique_ptr<aggregation>>(attr.value)->clone();
+      cudf::detail::aggregation_dispatcher(other_agg->kind, *this, *other_agg);
+      auto result = cache.get_result(values, *other_agg);
+      udf_input.emplace(std::move(other_agg), std::move(result));
+    }
+  }
+
+  auto output = (*udf_ptr)(udf_input, stream, mr);
+  CUDF_EXPECTS(std::holds_alternative<std::unique_ptr<column>>(output),
+               "Invalid output type from HOST_UDF groupby aggregation.");
```

Comment:
Perhaps the guts of this could be moved into a separate .cpp file (e.g. `group_host_udf.cpp`)? 
See some of the other groupby aggregations that do this.

Reply:
Except for this file, unfortunately.

Reply:
I've remove dependency from `host_udf_base` from all other places except for this one. At the end of this file, `aggregation_dispatcher` is called which needs to see all definitions of the `aggregation_result_functor` which are defined in this file.

Reply:
I tried to move this specialized function into a separate source file while leaving its forward declaration here so `aggregation_dispatcher` can successfully call it. However, this specialized function itself needs to call `aggregation_dispatcher` on all other aggregations! By moving it into a separate source file, it cannot see the other specializations thus unfortunately we need to leave it here together with all other specializations.

Reply:
I think that you could leave the `aggregate_result_functor::operator()<aggregation::HOST_UDF>` function definition here but move _most_ of the implementation into a separate file. Look at https://github.com/rapidsai/cudf/pull/17592/files#diff-a943ede169c709882f6825c90823e342e974bbb95ef1a5ebe1bea7f083caf811L573-L583 as example. Most of the implementation is actually in https://github.com/rapidsai/cudf/blob/branch-25.02/cpp/src/groupby/sort/group_merge_m2.cu

Reply:
I still cannot find any solution for it yet. The reason was already mentioned above:
 * The function at the end of this file needs to call `aggregation_dispatcher` which requires seeing the declaration/definition of all specializations.
 * This specialization also requires seeing the declaration/definition of all other specializations.

I can move this specialization out successfully if I add forward declaration for **all** other aggregations in its source file. That seems too much overkill for cutting of dependency for this source file, which is relatively small.

Reply:
Ok, that is fine. Thanks for investigating.